app/models/kpi.rb

```diff
@@ -0,0 +1,366 @@
+  }
+
+  CASE_CLOSURE_TYPE = {
+    name: 'case_closure',
+    calculations: [
+      'elapsed_time_from_opening_to_closure',
+      'elapsed_time_from_opening_to_closure_hight_risk',
+      'reason_for_case_closure',
+      'case_closure_rate'
+    ],
+    days_ranges: [
+      [0, 30],
+      [31, 92],
+      [93, 185],
+    ]
+  }
+
+  CLIENT_SATISFACTION_TYPE = {
+    name: 'client_satisfaction',
+    calculations: ['client_satisfaction']
+  }
+
+  OTHER_TYPE = {
+    name: 'other_kpi',
+    calculations: [
+      'case_load',
+      'supervisor_to_case_worker_ratio',
+      'caseworkers_active_on_gbvims_plus'
+    ],
+  }
+
+  TYPES = [
+    INTRODUCTION_AND_ENGAGEMENT_TYPE,
+    ASSESSMENT_TYPE,
+    CASE_ACTION_PLANNING_TYPE,
+    IMPLEMENT_THE_CASE_ACTION_PLAN_TYPE,
+    CASE_FOLLOW_UP_TYPE,
+    CASE_CLOSURE_TYPE,
+    CLIENT_SATISFACTION_TYPE,
+    OTHER_TYPE,
+  ]
+
+  NEEDED_DATE_RANGE = ['follow_up_meetings', 'completed_asssessment']
+
+  def self.minimum_reportable_fields
+    {
+      'string' => [
+        INTRODUCTION_AND_ENGAGEMENT_TYPE[:name],
+        ASSESSMENT_TYPE[:name],
+        CASE_ACTION_PLANNING_TYPE[:name],
+        IMPLEMENT_THE_CASE_ACTION_PLAN_TYPE[:name],
+        CASE_FOLLOW_UP_TYPE[:name],
+        CASE_CLOSURE_TYPE[:name],
+        CLIENT_SATISFACTION_TYPE[:name],
+        OTHER_TYPE[:name],
+      ]
+    }
+  end
+
+  def self.get_default_range_from(range)
+    range.each_with_object({}) do |(from, to), default|
+      default["#{from}-#{to}"] = 0
+    end
+  end
+
+  def self.to_percent value
+    "#{(value.to_f * 100).ceil(3)}%"
+  end
+
+  def self.format_result_for_reduced_view(view_result, range = [])
+    view_result.rows.each_with_object(get_default_range_from range) do |row, result|
+      result[row.key.kind_of?(Array) ? row.key[-1] : row.key] = row.value
+    end
+  end
+
+  def self.format_percents_from_rows(rows)
+    rows
+    .map {|key, result| [key, result.to_f / rows['_count'].to_f]} 
+    .to_h
+    .except '_count'
+  end
+
+  def self.get_all_open_cases_count(current_user, date_range = [])
+    all_active_cases_search = Child.search do
+      with :child_status, Record::STATUS_OPEN
+      with :owned_by_agency, current_user[:organization]
+      with(:registration_date).between(
+        Date.parse(date_range[0])..Date.parse(date_range[1])
+      ) if date_range[0].present? && date_range[1].present?
+    end
+    all_active_cases_search.total || 1
+  end
+
+  def self.generate_kpi(types, current_user, date_range)
+    kpi_calculation_results = Parallel.map(types, in_threads: types.length) do |kpi_type| 
+      [kpi_type, send(kpi_type, current_user, date_range)]
+    end
+    kpi_calculation_results.each_with_object({}) do |(kpi_type, calculation_result), kpi| 
+      group = TYPES.detect {|e| e[:calculations].include?(kpi_type) }
+      kpi[group[:name]] ||= {};
+      kpi[group[:name]][kpi_type] = calculation_result
+    end
+  end
+
+  def self.rate_of_report(current_user, date_range)
```

Comment:
Parameter "date_range" isn't used.  If it is necessary, use _ at the beginning
```suggestion
  def self.rate_of_report(current_user, _date_range)
```